### ISSUE TITLE
Dependency metadata serialization is broken for error caused by recursive dependency

### ISSUE THREAD

**Issue by jjohannes** [gradle/gradle#11844]
This is an edge case where a module depends on another version of itself through a dependency cycle, but that other version can not be retrieved, causing a resolution error that is not correctly serialized.

For example:

`org:lib:1.0 -> org:between:1.0 -> org:lib:1.1` (where `org:lib:1.1` is not found)

Test reproducing the issue:

https://github.com/gradle/gradle/blob/e079efb4ed7ae638373e2f66dc9e042e2ad2828f/subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/VersionConflictResolutionIntegrationTest.groovy#L1889-L1920

Error output:

```
> Could not resolve all dependencies for configuration ':runtimeClasspath'.
   > Problems reading data from Binary store in ...

Corrupt serialized resolution result. Cannot find selected module (4) for runtimeClasspath -> org:lib:1.0
```

**Comment by dansanduleac** [comment]
Just wanted to point out that we're still encountering this on some repos even with 6.2-rc-2.
In one case, this was still hit this error when resolving `testCompile` (which I know we shouldn't), but _only_ when resolving it in a loop with other configurations in that project being resolved first.

For instance, if we resolve `testCompile` directly (e.g. calling `dependencyInsight` on it) then it works fine.

```
Caused by: java.lang.RuntimeException: Problems reading data from Binary store in /private/var/folders/j3/6bm756yj08ng5ht8_5c9vt2j081g9_/T/gradle2102174969145096512.bin offset 1978926 exists? true
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.store.DefaultBinaryStore$SimpleBinaryData.read(DefaultBinaryStore.java:132)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.StreamingResolutionResultBuilder$RootFactory.lambda$create$0(StreamingResolutionResultBuilder.java:187)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.store.CachedStoreFactory$SimpleStore.load(CachedStoreFactory.java:101)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.StreamingResolutionResultBuilder$RootFactory.create(StreamingResolutionResultBuilder.java:185)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.StreamingResolutionResultBuilder$RootFactory.create(StreamingResolutionResultBuilder.java:157)
        at org.gradle.api.internal.artifacts.result.DefaultResolutionResult.getRoot(DefaultResolutionResult.java:47)
        at org.gradle.api.internal.artifacts.ivyservice.ErrorHandlingConfigurationResolver$ErrorHandlingResolutionResult.getRoot(ErrorHandlingConfigurationResolver.java:201)
        ... 14 more
Caused by: java.lang.IllegalStateException: Corrupt serialized resolution result. Cannot find selected module (486) for constraint platform-runtime -> com.palantir.ri:resource-identifier:1.0.1
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.DefaultResolutionResultBuilder.visitOutgoingEdges(DefaultResolutionResultBuilder.java:82)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.StreamingResolutionResultBuilder$RootFactory.deserialize(StreamingResolutionResultBuilder.java:237)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.store.DefaultBinaryStore$SimpleBinaryData.read(DefaultBinaryStore.java:130)
```

**Comment by abdullahizzuddiin** [comment]
Hello, I think I got same problem with this https://github.com/gradle/gradle/issues/8489

are there any hint to resolve this problem?

**Comment by jjohannes** [comment]
Hi @abdullahizzuddiin,

Different problems/bugs can cause a "Problems reading data from Binary store..." error. This issue is about a specific situation that can only appear if an already broken build with a specific kind of cyclic dependency is synced in the IDE (e.g. in Android Studio). That is, builds with the problem described in this issue will always fail, just with different errors depending on how they are utilized.

There is a chance that you discovered a different problem. Can you share some more information about your build and its dependencies? Maybe a build scan? Did the error appear in Gradle 6.2 or did it exist in previous versions as well?

**Comment by fdw** [comment]
I think we've been bitten by this bug; `gradle dependencies --stacktrace` ultimately results in this error: `java.lang.IllegalStateException: Corrupt serialized resolution result. Cannot find selected module (89) for constraint platform-compile`

Unfortunately, our setup is quite complicated. What information do you need?

**Comment by rolandkreuzer** [comment]
I've run into the same message and have a hunch it might be caused by Gradle Metadata.

```
> Could not resolve all dependencies for configuration ':testCompileClasspath'.
   > Problems reading data from Binary store in C:\Users\rk\AppData\Local\Temp\gradle12148494386777462623.bin offset 175231 exists? true

Caused by: java.lang.IllegalStateException: Corrupt serialized resolution result. Cannot find selected module (654) for apiElements -> org.assertj:assertj-core
```

It seems the issue only occures since both the project defining the Platform and the project consuming the platform use Gradle 6 ( was able to reproduce in latest 6.5.1 )

I was able to workaround the issue by adding ignoreGradleMetadataRedirection in repository declaration.

```
maven {
    url 'https://internalmvnrepo'
    metadataSources {
        ignoreGradleMetadataRedirection()
        mavenPom()
    }
}
```

An alternative way to workaround is to downgrade the project publishing the Platform to Gradle 5 and therefore only publishing a MVN POM instead of both POM and Gradle Metadata.

Therefore I'd think the issue has to do with Gradle Metadata modules and/or a mixture with Non-Gradle-Metadata platforms.

I won't be able to share our complete platform definition, but it can be outlined as following:

```
+--- all-dependencies-platform
|    +--- internal-datatypes-platform (constraints only)
|    +--- external-dependencies-platform (constraints and platform refs)
|    +--- spring-dependencies-platform (constraints and platform refs)
|         +--- org.springframework.boot:spring-boot-dependencies
```

A shortened version of the setup would look something like

```
javaPlatform {
    allowDependencies()
}

dependencies {

    api platform('org.springframework.boot:spring-boot-dependencies:2.2.8.RELEASE')

    constraints {
        api ('org.assertj:assertj-core:3.16.1')
    }
}
```

The assertj dependency is included in both an internal constraints definition (with Gradle Metadata) and indirectly by the imported Spring Boot BOM (without Gradle Metadata).

Hope that helps.

**Comment by donat** [comment]
It seems that this has been fixed in Gradle 7.3 https://github.com/gradle/gradle/commit/ef3271927c0f2ea2cfecf27ed199b7178439f322

Everyone following this thread: can you please confirm that updating to the [latest 7.3 RC/GA](https://github.com/gradle/gradle/releases) fixes your use-case?

**Comment by donat** [comment]
Closing as fixed.
